plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -277,51 +266,51 @@ annotate = flip evalState 0 . flip runReaderT [] . go
-  Term name uni fun (Path, ann) ->
-  -- | Here, the value of the inner map not only contains the count, but also contains
+countOccs
+  :: forall name uni fun ann
+   . (Hashable (Term name uni fun ()), ToBuiltinMeaning uni fun)
+  => BuiltinSemanticsVariant fun
+  -> Term name uni fun (Path, ann)
+  -> HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)]
+  -- ^ Here, the value of the inner map not only contains the count, but also contains
   -- the annotated term, corresponding to the term that is the key of the outer map.
   -- The annotated terms need to be recorded since they will be used for substitution.
-  HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)]
-countOccs arityInfo = foldrOf termSubtermsDeep addToMap Map.empty
-  where
-    addToMap ::
-      Term name uni fun (Path, ann) ->
-      HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)] ->
-      HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)]
-    addToMap t0
-      -- We don't consider work-free terms for CSE, because doing so may or may not
-      -- have a size benefit, but certainly doesn't have any cost benefit (the cost
-      -- will in fact be slightly higher due to the additional application).
-      --
-      -- `isWorkFree` currently doesn't check whether a builtin application is saturated,
-      -- or whether an term is the (possibly repeated) forcing of a builtin (which should
-      -- be workfree), so we check it separately.
-      | isWorkFree t0 || not (isBuiltinSaturated t0) || isForcingBuiltin t0 = id
-      | otherwise =
-          Map.alter
-            ( \case
-                Nothing -> Just $ [(path, t0, 1)]
-                Just paths -> Just $ combinePaths t0 path paths
-            )
-            t
-      where
-        t = void t0
-        path = fst (termAnn t0)
-
-    isBuiltinSaturated = \case
-      t@Apply{}
-        | (Builtin _ fun, args) <- splitApplication t ->
-            length args >= Map.findWithDefault 0 fun arityInfo
-      _ -> True
-
-    isForcingBuiltin = \case
-      Builtin{} -> True
-      Force _ t -> isForcingBuiltin t
-      _ -> False
+countOccs builtinSemanticsVariant = foldrOf termSubtermsDeep addToMap Map.empty
+ where
+  addToMap
+    :: Term name uni fun (Path, ann)
+    -> HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)]
+    -> HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)]
+  addToMap t0
+    -- We don't consider work-free terms for CSE, because doing so may or may not
+    -- have a size benefit, but certainly doesn't have any cost benefit (the cost
+    -- will in fact be slightly higher due to the additional application).
+    --
+    -- `isWorkFree` currently doesn't check whether a builtin application is saturated,
+    -- or whether an term is the (possibly repeated) forcing of a builtin (which should
+    -- be workfree), so we check it separately.
```

Comment:
I think this comment is now inaccurate, and you can get rid of the extra conditions?